### PR TITLE
[ADD] l10n_my_edi: Malaysian E-Invoicing

### PR DESCRIPTION
**Pull request by vin-odoo**
Add support for connecting to the MyInvois API, in order to send and receive E-Invoices.

see https://github.com/odoo/iap-apps/pull/849

Task id # 3953659

### CODE REVIEW

**Review by rugo-odoo — CHANGES_REQUESTED** [review]
Nice job ! Here is the first review.

File: odoo/addons/base/data/res.country.state.csv
Comment (vin-odoo) [reply odoo/addons/base/data/res.country.state.csv:1]: in 17.0? :smile: 

I think it'll be updated during migration, or when base is -u
For the other cases I added a fallback in the code to ensure it'll work with old values too.

File: addons/account_edi_proxy_client/views/account_edi_proxy_user_views.xml
Comment (vin-odoo) [reply addons/account_edi_proxy_client/views/account_edi_proxy_user_views.xml:62]: Basically, it shows all records (active and inactive). It's a debug view, and it makes it easier IMO to understand what is there.

File: addons/l10n_my_edi/models/account_edi_proxy_user.py
Comment (vin-odoo) [reply addons/l10n_my_edi/models/account_edi_proxy_user.py:30]: Not really, we don't have these yet :sweat: 
I would prefer to finish getting my hand on a digital certificate before setting up the sandbox environment (and prod later).

I will think about maybe seeing with IAP to setup the test endpoint.

File: addons/l10n_my_edi/models/account_edi_xml_ubl_my.py
Comment (vin-odoo) [reply addons/l10n_my_edi/models/account_edi_xml_ubl_my.py:237]: No, I'm not enforcing a format in the field. I could check when generating the ubl field, do you think it's better.
Technically, we just put it in the export as is.
I guess we could split on ; and ensure that we have at most two values

File: addons/l10n_my_edi/models/account_move.py
Comment (vin-odoo) [reply addons/l10n_my_edi/models/account_move.py:143]: We do, otherwise the button will not show on invoices without state (bills, ... that you wouldn't send to the government)

File: addons/l10n_my_edi/models/account_move.py
Comment (vin-odoo) [reply addons/l10n_my_edi/models/account_move.py:222]: Yes but no, as we give a message param to the method call we need the var to be assigned.
It's either in the else, or before the if but it's needed ^^

File: addons/l10n_my_edi/models/account_move.py
Comment (vin-odoo) [reply addons/l10n_my_edi/models/account_move.py:365]: It's the result of the api call :smile: 
If you have another idea let me know, but to me it's clear enough

File: addons/l10n_my_edi/models/account_move.py
Comment (vin-odoo) [reply addons/l10n_my_edi/models/account_move.py:385]: yes, I probably used new_status at some point and remove that later :+1:

File: addons/l10n_my_edi/models/account_move.py
Comment (vin-odoo) [reply addons/l10n_my_edi/models/account_move.py:416]: Forgot what I was doing with that :joy: 
I'll remove it for now

File: addons/l10n_my_edi/models/account_move.py
Comment (vin-odoo) [reply addons/l10n_my_edi/models/account_move.py:495]: The status can change up to 72h after sending it as you saw.
The 74 is to take into account that the cron may runs once an hour, and we want to ensure that we get the new status if the invoice was updated right before the 72-hour mark.
Clarified it a bit more in the docstring

File: addons/l10n_my_edi/models/account_move.py
Comment (vin-odoo) [reply addons/l10n_my_edi/models/account_move.py:738]: :thinking: But we don't always receive a JSON, that's the point.
What we receive is either a JSON or a XML depending on the method used by the supplier.
Here we avoid repeating the check:

JSON:
    - We convert to XML
    - We check the validity of it
Any other
    - We skip the first step and check the validity of it

File: addons/l10n_my_edi/models/account_move.py
Comment (vin-odoo) [reply addons/l10n_my_edi/models/account_move.py:749]: We can't, same reason as above :smile: 
We can discuss that if needed.

File: addons/l10n_my_edi/wizard/account_move_send.py
Comment (vin-odoo) [reply addons/l10n_my_edi/wizard/account_move_send.py:139]: Sorry, this is a bit :spaghetti: 
_l10n_my_edi_generate_myinvois_xml will populate 'myinvois_attachments' if it went as expected, so this checks avoid trying to decode the XML if it failed for any configuration issue

File: addons/l10n_my_edi/models/account_edi_xml_ubl_my.py
Comment (rugo-odoo) [reply addons/l10n_my_edi/models/account_edi_xml_ubl_my.py:237]: I think it will be better. It's weird to mention it on the docstring but not doing something about it. Also it a quick/easy check.

**Review by rugo-odoo — CHANGES_REQUESTED** [review]

File: addons/l10n_my_edi/__manifest__.py
Comment (antoine162) [inline addons/l10n_my_edi/__manifest__.py:28]: Should we make it auto-install? (according to JCO, our dear enlightened CEO has requsted in the past for EDIs to be auto-installed)
Note that if we do that, we should check that this doesn't break anything in the 16.0 -> 17.0 migration. But as far as I see, it shouldn't break anything.

File: addons/l10n_my_edi/models/account_move.py
Comment (antoine162) [inline addons/l10n_my_edi/models/account_move.py:859]: Isn't this going to give a weird error to the user?
Say, they update the status, and then get a lock date error.
Maybe this is the lesser of two evils (if locked entries really really must not be changed)
but in my view this is going to be quite confusing for users.

**Review by antoine162 — COMMENTED** [review]
To me, this is really really good. I saw you managed to avoid the usual pitfalls - double sending, serialization errors on flush. The error handling seems legit.

Just one question regarding cancelled documents, especially. I see they're expected to be re-sent. Is there, more generally, a requirement on invoice sequencing?

Happy to r+ when you think this is ready.

File: addons/l10n_my_edi/__manifest__.py
Comment (vin-odoo) [reply addons/l10n_my_edi/__manifest__.py:28]: I usually prefer not to, but if you think it is better I can.
We just released an edi in Vietnam and I wasn't asked to make it auto install :thinking:

File: addons/l10n_my_edi/models/account_move.py
Comment (vin-odoo) [reply addons/l10n_my_edi/models/account_move.py:859]: It could be.
This is a mistake I made in VN too anyway; we don't want to raise here.
It would cause losing some important information in case we are cancelling using the wizard, it succeeds on the platform but fails in Odoo for any reason.

In practice, these should never fail as there is a 72-hour window for these changes and it is unlikely that you lock a period so close to its end date.

I added a try and catch around the lines; and if the cancellation fails it will post a message instead to warn about it.

File: addons/l10n_my_edi/__manifest__.py
Comment (antoine162) [reply addons/l10n_my_edi/__manifest__.py:28]: For me, it's fine to leave it as it is now :+1:

File: addons/l10n_my_edi/models/account_move.py
Comment (antoine162) [inline addons/l10n_my_edi/models/account_move.py:465]: - `invoices_per_company.keys()` is already a set-like object
- You could directly iterate on `for company, company_invoices in invoices_per_company.items()`

File: addons/l10n_my_edi/models/account_move.py
Comment (antoine162) [inline addons/l10n_my_edi/models/account_move.py:525]: 1 second or 3 seconds?

File: addons/l10n_my_edi/models/account_move.py
Comment (antoine162) [inline addons/l10n_my_edi/models/account_move.py:569]: What if the proxy user is deactivated? should we still use it to send/retrieve invoices?

**Review by antoine162 — COMMENTED** [review]
Looks great to me, just a few comments/questions but apart from that it's good to merge imo :smile:

File: addons/l10n_my_edi/models/account_edi_proxy_user.py
Comment (antoine162) [inline addons/l10n_my_edi/models/account_edi_proxy_user.py:36]:
```suggestion
            'prod': 'http://l10n-my-edi.api.odoo.com',
            'test': self.env['ir.config.parameter'].get_param('l10n_my_edi_test_server_url', 'http://l10n-my-edi.test.odoo.com'),
```
Something like this

File: addons/l10n_my_edi/models/account_move.py
Comment (vin-odoo) [reply addons/l10n_my_edi/models/account_move.py:525]: Sorry, 1s indeed.
The API asks to not repeat the call for the same submission more than once every 3-5s
Here we are getting different submissions and considering there will be a delay while waiting for their response/... 1s seems enough.

File: addons/l10n_my_edi/models/account_move.py
Comment (vin-odoo) [reply addons/l10n_my_edi/models/account_move.py:569]: I believe company.l10n_my_edi_proxy_user_id will return an empty recordset if the user is deactivated, and thus it will not show the option to send in the send & print/will raise.
It may not be perfect, but the only use case where this should happen would be duplicate databases where we don't want it to work anyway.

Comment (antoine162) [reply addons/l10n_my_edi/models/account_move.py:569]: You're right about that! Never mind :smile:

**Review by antoine162 — APPROVED** [review]
LGTM, this looks like one of the best EDIs I've ever seen :smile:
@robodoo r+

**Comment by antoine162** [comment]
@robodoo rebase-ff

**Comment by antoine162** [comment]
(actually, let's wait for the IAP branch to be reviewed first before merging)
@robodoo r-

**Comment by antoine162** [comment]
@robodoo r+

**Pull request closed without merging**
